File: Data Structures & Algorithms/island-perimeter/submission-1.py
class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        directions = [(0,1), (1,0), (0,-1), (-1,0)]

        def dfs(i, j):

            # Out of bounds contributes perimeter
            if i < 0 or j < 0 or i >= m or j >= n:
                return 1

            # Water contributes perimeter
            if grid[i][j] == 0:
                return 1

            # Already visited
            if grid[i][j] == -1:
                return 0

            # Mark visited
            grid[i][j] = -1

            perimeter = 0

            for dx, dy in directions:
                perimeter += dfs(i + dx, j + dy)

            return perimeter

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    return dfs(i, j)

        return 0

File: Data Structures & Algorithms/count-number-of-islands/submission-1.go
func numIslands(grid [][]byte) int {
	m := len(grid)
	n := len(grid[0])
    count := 0

    traces := [][]int{
		{0, 1},
		{1, 0},
		{0, -1},
		{-1, 0},
	}

    var dfs func(x, y int) 

    dfs = func(x, y int) {
        if x < 0 || y < 0 || x >= m || y >= n || grid[x][y] == '0' {
			return 
		}
	
		grid[x][y] = '0'
        for _, value := range traces {
			xi, yj := x + value[0], y + value[1]
			dfs(xi, yj)
		}
    }


    for i := range m{
        for j := range n {
            if grid[i][j] == '1' {
                count += 1
                dfs(i, j)
            }
        }
    }

    return count
}
